### PR TITLE
[Draft] Join, Sync, Heartbeat during Migration

### PR DESCRIPTION
**Pull request by dongnuo123**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java [hunk 1]
```diff
@@ -3467,6 +3476,358 @@ public void maybeDeleteGroup(String groupId, List<Record> records) {
...
+                // TODO: need a list of all available server assignors
+                if (UniformAssignor.UNIFORM_ASSIGNOR_NAME.equals(protocol.name())
+                    || RangeAssignor.RANGE_ASSIGNOR_NAME.equals(protocol.name())) {
+                    return protocol;
+                }
+            }
+        }
+        throw new FencedMemberEpochException("The JoinGroup request doesn't have a matched generation id from a " +
+            "protocol supported by the server assignors with the epoch of the member known by the group coordinator (" +
+            member.memberEpoch() + ").");
```
Comment (dongnuo123) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: not supported by the old protocol

Comment (dongnuo123) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: maybe handle the case where generation id doesn't match the member epoch.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java [hunk 2]
```diff
@@ -3467,6 +3476,358 @@ public void maybeDeleteGroup(String groupId, List<Record> records) {
...
+        assignedPartitions.entrySet().forEach(item -> {
+            TopicImage topicImage = metadataImage.topics().getTopic(item.getKey());
+            if (topicImage == null) {
+                throw INVALID_TOPIC_EXCEPTION.exception("Can't find the topic id of topic " + item.getKey() + ".");
+            }
+            partitions.addAll(item.getValue().stream().map(partition ->
+                new TopicPartition(topicImage.name(), partition)).collect(Collectors.toList()));
+        });
+
+        final ByteBuffer byteBuffer = ConsumerProtocol.serializeAssignment(new ConsumerPartitionAssignor.Assignment(partitions)); // TODO: assignment version?
```
Comment (dongnuo123) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: deserializeVersion

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java [hunk 3]
```diff
@@ -3467,6 +3476,358 @@ public void maybeDeleteGroup(String groupId, List<Record> records) {
...
+            responseFuture.complete(new SyncGroupResponseData()
+                .setErrorCode(Errors.forException(t).code())
+            );
+            return EMPTY_RESULT;
+        }
+
+        // TODO:
+        // 1) member id exists, instance id exists and matches member id
+        // 2) generation id and protocol match
+        // 3) group state is not empty or dead?
```
Comment (dongnuo123) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: accept assigning state

Comment (dongnuo123) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: check if empty or dead is possible if 1) 2) passes

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java [hunk 4]
```diff
@@ -3467,6 +3476,358 @@ public void maybeDeleteGroup(String groupId, List<Record> records) {
...
+
+        scheduleConsumerGroupSessionTimeout(groupId, memberId);
+
+        if (group.groupEpoch() > member.memberEpoch() ||
+            !member.partitionsPendingRevocation().isEmpty() ||
+            !member.partitionsPendingAssignment().isEmpty()) {
+            new HeartbeatResponseData().setErrorCode(Errors.REBALANCE_IN_PROGRESS.code());
```
Comment (dongnuo123) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:3846]: set join group timeout

**Comment by dajac** [comment]
I think that we don't need this anymore. Closing.

**Pull request closed without merging**